lib/OpenQA/Task/Bug/Limit.pm

```diff
@@ -0,0 +1,84 @@
+# Copyright (C) 2018-2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::Task::Bug::Limit;
+use Mojo::Base 'Mojolicious::Plugin';
+
+use File::Spec::Functions 'catfile';
+use File::Basename qw(basename dirname);
+use OpenQA::Utils qw(:DEFAULT imagesdir);
+use Data::Dump 'pp';
+use Try::Tiny;
+
+sub register {
+    my ($self, $app) = @_;
+    $app->minion->add_task(limit_bugs => sub { _limit($app, @_) });
+}
+
+sub _limit {
+    my ($app, $job) = @_;
+
+    # prevent multiple limit_bugs tasks to run in parallel
+    return $job->finish('Previous limit_bugs job is still active')
+      unless my $guard = $app->minion->guard('limit_bugs_task', 86400);
+
+    # prevent multiple limit_* tasks to run in parallel
+    return $job->retry({delay => 60})
+      unless my $limit_guard = $app->minion->guard('limit_tasks', 86400);
+
+    try {
+        # cleanup entries in the bug table that are not referenced from any comments
+        my $bugrefs = bugrefcount($app);
+        my $bugs    = $app->schema->resultset("Bugs")->search();
+        my %cleaned;
+        for my $bug ($bugs->all) {
+            if (!defined $bugrefs->{$bug->bugid}) {
+                $bug->delete;
+                $cleaned{$bug->id} = $bug->bugid;
+            }
+        }
+        $app->emit_event('openqa_bugs_cleaned', {deleted => scalar(keys(%cleaned))});
+        $job->note(bugs_cleaned => pp(%cleaned));
+    }
+    catch {
+        my $error = $_;
+        $job->fail($error);
+    };
+}
+
+=over 4
+
+=item bugrefcount()
+
+Return a hashref of all bugs referenced by job comments with the number of references.
+
+=back
+
+=cut
+
+sub bugrefcount {
```

Comment:
This should be a method of the resultset. I'd also call it `compute_bugref_count` to make it clear that it really computes the value on the fly (and not just returns a previously computed value). You can get the app via `OpenQA::App->singleton` from there.

Reply:
I don't see why it would have to access the application at all. Wouldn't it just be a `Comments` resultset method?

Reply:
So which file would that be?
And how would I call this then?

Reply:
`lib/OpenQA/Schema/ResultSet/Comments.pm` and `OpenQA::Schema->singleton->resultset('Comments')->bugrefcount`.

Reply:
Or `$app->schema->resultset('Comments')->bugrefcount`, depending on context.

Reply:
1. `lib/OpenQA/Schema/ResultSet/Bugs.pm`
2. `$app->schema->resultset('Comments')->compute_bugref_count`